From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: libalpm64 <159192189+libalpm64@users.noreply.github.com>
Date: Sun, 5 Jul 2026 00:00:00 +0000
Subject: [PATCH] Speleothem: Remove damageable+stackable item validation for 32k
 items


diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 0b016af71414cf6527839b178996115129e43323..20c4d8ae16d9c8af14321188dbf35e8234eb0891 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -286,7 +286,7 @@ public final class ItemStack implements DataComponentHolder, ItemInstance, Chang
     }
 
     private static DataResult<?> validateComponents(final DataComponentMap components) {
-        if (components.has(DataComponents.MAX_DAMAGE) && components.getOrDefault(DataComponents.MAX_STACK_SIZE, 1) > 1) {
+        if (false && components.has(DataComponents.MAX_DAMAGE) && components.getOrDefault(DataComponents.MAX_STACK_SIZE, 1) > 1) {
             return DataResult.error(() -> "Item cannot be both damageable and stackable");
         }
 
