From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: libalpm64 <159192189+libalpm64@users.noreply.github.com>
Date: Thu, 6 Aug 2026 02:18:54 -0400
Subject: [PATCH] Speleothem: Prevent oversized inventory items from
 disconnecting players


diff --git a/net/minecraft/network/protocol/game/ClientboundContainerSetSlotPacket.java b/net/minecraft/network/protocol/game/ClientboundContainerSetSlotPacket.java
index 1879ed63e5bf1d39756e69b1943a72c0769fcfe7..dd9983d35e592061b9c25af4fe6a37dd1e8c968e 100644
--- a/net/minecraft/network/protocol/game/ClientboundContainerSetSlotPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundContainerSetSlotPacket.java
@@ -39,3 +39,16 @@ public class ClientboundContainerSetSlotPacket implements Packet<ClientGamePacke
+    // Speleothem start - oversized item fallback
+    @Override
+    public boolean hasLargePacketFallback() {
+        return !this.itemStack.isEmpty();
+    }
+
+    @Override
+    public boolean packetTooLarge(final net.minecraft.network.Connection connection) {
+        connection.send(new ClientboundContainerSetSlotPacket(this.containerId, this.stateId, this.slot, ItemStack.EMPTY));
+        return true;
+    }
+    // Speleothem end - oversized item fallback
+
     @Override
     public PacketType<ClientboundContainerSetSlotPacket> type() {
         return GamePacketTypes.CLIENTBOUND_CONTAINER_SET_SLOT;
